src/pages/p_oauth.php

```diff
@@ -242,13 +243,31 @@ private function instance_response($authi, $tok, $tokdata) {
         if (isset($jid->orcid) && is_string($jid->orcid)) {
             $reg["orcid"] = $jid->orcid;
         }
+        if (isset($jid->affiliation) && is_string($jid->affiliation)) {
+            $reg["affiliation"] = $jid->affiliation;
+        }
         $info = LoginHelper::check_external_login(Contact::make_keyed($this->conf, $reg));
         if (!$info["ok"]) {
             LoginHelper::login_error($this->conf, $jid->email, $info, null);
             throw new Redirection($tokdata->site_uri);
         }
 
         $user = $info["user"];
+        error_log(var_export($authi, true)."\n");
+        if (isset($jid->groups) && isset($authi->group_mappings)) {
+            foreach ($authi->group_mappings as $group => $role) {
+                if (in_array($group, $jid->groups, true)) {
+                    error_log("setting: ".$group." ".constant("Contact::".$role)."\n");
+                    $user_roles = $user->roles | constant("Contact::".$role);
+                    $user->save_roles($user_roles, $user);
+                } else {
+                    error_log("removing: ".$group." ".constant("Contact::".$role)."\n");
```

Comment:
See above

Reply:
Is it actually a good idea to *remove* database roles based on OAuth information? Seems odd.

Reply:
We can make this configurable; However, if you have, e.g., centralized authentication and authorization for a project, you want to ensure that permissions are removed when done so centrally.

What might make sense is having a switch for it. Let me take a look.